haystack/eval.py

```diff
@@ -1,7 +1,217 @@
 from typing import List, Tuple, Dict, Any
+import logging
 
 from haystack import MultiLabel
 
+from farm.evaluation.squad_evaluation import compute_f1 as calculate_f1_str
+from farm.evaluation.squad_evaluation import compute_exact as calculate_em_str
+
+logger = logging.getLogger(__name__)
+
+
+class EvalRetriever:
+    def __init__(self, debug=False, open_domain=True):
+        self.outgoing_edges = 1
+        self.init_counts()
+        self.no_answer_warning = False
+        self.debug = debug
+        self.log = []
+        self.open_domain = open_domain
+
+    def init_counts(self):
+        self.correct_retrieval_count = 0
+        self.query_count = 0
+        self.has_answer_count = 0
+        self.has_answer_correct = 0
+        self.has_answer_recall = 0
+        self.no_answer_count = 0
+        self.recall = 0.0
+
+    def run(self, documents, labels: dict, **kwargs):
+        self.query_count += 1
+        retriever_labels = labels["retriever"]
+        # TODO retriever_labels is currently a Multilabel object but should eventually be a RetrieverLabel object
+        # If there is a no_answer annotation in the labels
+        if "" in retriever_labels.multiple_answers:
```

Comment:
Couldnt this cuase bugs? Cant we test for no answer in a better way? If there is a mix of ["", "answer1", "answer2"] we should still consider it as text answer.

Reply:
Sure, we now check `MultiLabel.no_answer` to determine if a sample is a `no_answer` sample